### ISSUE TITLE
Porting to Crossbeam

### ISSUE THREAD

**Issue by jeehoonkang** [spacejam/sled#116]
Hi!

I would like to report my experience in porting `rsdb` to the new crossbeam.  Here is the implementation: https://github.com/jeehoonkang/rsdb/tree/crossbeam-epoch  The diff is not that much: https://github.com/jeehoonkang/rsdb/commit/194a41209a28774da252843437d16327bd188846

I think the performance is comparable to the `master` branch.  Here are some benchmark results I measured with `cargo build --bin=stress --features=stress,zstd; hack/flamerun.sh ./target/debug/stress --duration 5` in a x86-64 machine (Intel(R) Xeon(R) CPU E3-1245 v3 @ 3.40GHz, 32GB RAM, Manjaro Linux):

crossbeam-epoch:
```
did 1992 ops                                        
did 1330 ops                                             
did 1825 ops                                                                          
did 1250 ops                                                                          
did 1787 ops                               
did 8184 total ops in 5 seconds. 1636 ops/s 
```

master:
```
did 2476 ops                               
did 858 ops                                
did 741 ops                                
did 1394 ops                               
did 1683 ops                               
did 7152 total ops in 5 seconds. 1430 ops/s
```

Of course, this is one-time measurement, and more systematic approach should be taken for a conclusive argument.  Here are some flamegraphs: [flamegraphs.zip](https://github.com/spacejam/rsdb/files/1279393/flamegraphs.zip)

Once `crossbeam-epoch` is published in `crates.io`, I will make a PR based on this branch.  Thanks!

**Comment by spacejam** [comment]
@jeehoonkang thank you very much for the flamegraphs! This is super helpful for tracking down this (really huge) performance bug. I see that punching holes in the file with fallocate is taking a ton of time on your system. Could you help me narrow down possible causes with some info:

1. what drive was used for storage?
1. are you running on bare metal, inside a VM, or inside a container? 
1. what filesystem are you using?
1. what kernel version do you have? 
1. were all 8 threads on your 4 cores relatively quiet other than this stress test? if another workload was running, was it touching the disk very much?

Thank you so much for this info! It is extremely helpful for helping understand the properties at play.

I am really excited about `crossbeam-epoch`! This kind of system really made handling garbage in this database fun :) Looking forward to the crate being published :)

**Comment by jeehoonkang** [comment]
I'm glad it's helpful!  I'm using INTEL SSDSC2BB30 279.5G, formatted ext4, on a bare metal machine running kernel 4.12.10-1-MANJARO.  Other workloads were running, e.g. emacs, firefox, and zsh+tmux.  But I don't think they incur significant perturbation, especially on disk performance.

**Comment by spacejam** [comment]
@jeehoonkang very interesting! This tells me that there is a major flaw in my assumption that fallocate should work easily on SSD's with modern linux kernels and ext4. I may shift gears immediately to implement something more along the lines of [LFS](https://people.eecs.berkeley.edu/~brewer/cs262/LFS.pdf) which is necessary anyway for supporting older kernels, but I was hoping to put it off until later haha...

This is a great thing to learn before I release an alpha/beta!

**Comment by jeehoonkang** [comment]
Progress report!

Recently (actually, today) we released [crossbeam-epoch 0.1.0](https://crates.io/crates/crossbeam-epoch).  Naturally I ported sled to it: https://github.com/jeehoonkang/rsdb/commit/4fdbcc093bb327ffa62042fb1c09922d105d217e

While it compiles, `cargo test` fails with an unknown error.  I'm looking into the memory/thread sanitizer results.
